rllib/utils/replay_buffers/mixin_multi_agent_replay_buffer.py

```diff
@@ -0,0 +1,293 @@
+                    return int(np.ceil(product))
+                else:
+                    return int(np.floor(product))
+
+            max_num_new = round_up_or_down(num_items, 1 - self.replay_ratio)
+            # if num_samples * self.replay_ratio is not not round,
+            # we need one more sample with a probability of
+            # (num_items*self.replay_ratio)%1
+
+            _buffer = self.replay_buffers[_policy_id]
+            output_batches = self.last_added_batches[_policy_id][:max_num_new]
+            self.last_added_batches[_policy_id] = self.last_added_batches[
+                                                      _policy_id][max_num_new:]
+
+            # No replay desired
+            if self.replay_ratio == 0.0:
+                return SampleBatch.concat_samples(output_batches)
+            # Only replay desired
+            elif self.replay_ratio == 1.0:
+                return _buffer.sample(num_items,
+                                      beta=self.prioritized_replay_beta)
+
+            if np.isclose(max_num_new, num_items * (1 - self.replay_ratio)):
+                # The optimal case, we can mix in a round number of old
+                # samples on average
+                num_old = num_items - max_num_new
+            else:
+                num_new = len(output_batches)
+                # We never want to return more elements than num_items
+                num_old = min(num_items - max_num_new,
+                              round_up_or_down(num_new,
+                                               self.replay_ratio
+                                               / (1 - self.replay_ratio)))
+
+            output_batches.append(_buffer.sample(num_old,
+                                                 self.prioritized_replay_beta))
+            return SampleBatch.concat_samples(output_batches)
+
+        def check_buffer_is_ready(_policy_id):
+            if (len(self.replay_buffers[policy_id]) == 0) or \
+                (len(self.last_added_batches[_policy_id]) == 0 and
+                 self.replay_ratio < 1.0):
+                return False
+            return True
+
+        with self.replay_timer:
+            if self.replay_mode == "lockstep":
+                assert (
+                    policy_id is None
+                ), "`policy_id` specifier not allowed in `locksetp` mode!"
+                if check_buffer_is_ready(_ALL_POLICIES):
+                    return mix_batches(_ALL_POLICIES)
+            elif policy_id is not None:
+                if check_buffer_is_ready(policy_id):
+                    return mix_batches(policy_id)
+            else:
+                samples = {}
+                for policy_id, replay_buffer in self.replay_buffers.items():
+                    if check_buffer_is_ready(policy_id):
+                        samples[policy_id] = mix_batches(policy_id)
+                return MultiAgentBatch(samples, self.replay_batch_size)
```

Comment:
here though, we will always return a MultiAgentBatch, even if we didn't collect anything.

Reply:
True. Ok to return None if not ready, but should be consistent between these cases, no?

Reply:
I have changed this to always returning a MA Batch, but empty if the buffer is empty.
Hope this works for you.